### PR TITLE
Same as PR-91

### PR DESCRIPTION
**Pull request by lovesh**
Because  https://github.com/docknetwork/dock-substrate/pull/91  couldn't be reopened

### CODE REVIEW

**Review by bddap — COMMENTED** [review]
Since only council members can vote the form of governance introduced in this PR not democracy. It's not representative democracy either since council members aren't elected by the populous.

Since council members can, without notable restriction, slash people for making or seconding any proposal, it's not safe for token holders to make proposals, especially controversial ones.

If token holders were the only source of proposals, that would grant them some power (albeit with the risk of slashing), but that power is undermined by the councils ability to make external proposals. The council need not vote yes on any public proposal because they can propose and vote on their own.

It took a long time to decipher how the desired feature set was implemented. I've noted in the review comments where and what documentation would have been helpful. I've also noted a couple things that would have helped communicate intent (see `simple_democracy::vote()` and `impl pallet_democracy::Trait for Runtime {`).

File: pallets/simple_democracy/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        type Error = Error<T>;
+
+        fn deposit_event() = default;
+
+        /// Proxy function to forked democracy pallet's `propose`
+        // TODO: Fix weight
+        #[weight = 0]
+        fn propose(origin, proposal_hash: T::Hash, #[compact] value: BalanceOf<T>) {
```
[the remark this answers is not in the document]
Comment (lovesh) [reply pallets/simple_democracy/src/lib.rs:65]: We can show a note on the extrinsics page for that.

File: runtime/src/lib.rs [hunk 2]
```diff
@@ -175,9 +183,42 @@ parameter_types! {
...
+            | Call::Utility(_)
+            | Call::PoAModule(_)
+            | Call::DIDModule(_)
+            | Call::Revoke(_)
+            | Call::BlobStore(_)
+            | Call::Master(_)
+            | Call::Sudo(_)
+            | Call::MigrationModule(_)
+            | Call::RandomnessCollectiveFlip(_) => true,
+            _ => false,
```
[the remark this answers is not in the document]
Comment (lovesh) [reply runtime/src/lib.rs:214]: Good catch. I have moved that filter closer to `construct_runtime` and added comment on both the filter and `construct_runtime`

File: pallets/simple_democracy/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,323 @@
+//! This is a facade over Substrate's democracy pallet offering similar functionality with a few differences.
+//! This pallet implements a "simple majority" governance where only Council can vote but anyone can
+//! propose by locking a fixed amount of tokens, `PublicProposalDeposit`. Similar to Substrate's governance pallet, both Council and
+//! general public take turns proposing, and a proposal follows "propose" -> "referendum" -> "enact" sequence.
+//! General public's proposals can be "seconded" by other token holder.
+//! Technical committee can fast track proposal made by Council member.
+//! A simple majority of Council members' votes (>50%) are needed to accept an ongoing referendum.
+//! Council members can cancel a proposal made by the public.
+//! Setting up of Council and Technical committee can be seen in the pallet's tests.
+
```
[the remark this answers is not in the document]
Comment (lovesh) [reply pallets/simple_democracy/src/lib.rs:10]: > This module implicitly relies on the fact that ...

Will mention that in comment.

File: pallets/simple_democracy/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,323 @@
+//! This is a facade over Substrate's democracy pallet offering similar functionality with a few differences.
+//! This pallet implements a "simple majority" governance where only Council can vote but anyone can
+//! propose by locking a fixed amount of tokens, `PublicProposalDeposit`. Similar to Substrate's governance pallet, both Council and
+//! general public take turns proposing, and a proposal follows "propose" -> "referendum" -> "enact" sequence.
+//! General public's proposals can be "seconded" by other token holder.
+//! Technical committee can fast track proposal made by Council member.
+//! A simple majority of Council members' votes (>50%) are needed to accept an ongoing referendum.
+//! Council members can cancel a proposal made by the public.
+//! Setting up of Council and Technical committee can be seen in the pallet's tests.
```
[the remark this answers is not in the document]
Comment (lovesh) [reply pallets/simple_democracy/src/lib.rs:9]: I will reword this.

File: pallets/simple_democracy/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+
+        /// Proxy function to forked democracy pallet's `enact_proposal`
+        #[weight = T::MaximumBlockWeight::get()]
+        fn enact_proposal(origin, proposal_hash: T::Hash, index: ReferendumIndex) -> DispatchResult {
+            <pallet_democracy::Module<T>>::enact_proposal(origin, proposal_hash, index)
+        }
+
+        /// Proxy function to forked democracy pallet's `note_preimage`
+        // TODO: Fix weight
+        #[weight = 0]
```
[the remark this answers is not in the document]
Comment (lovesh) [reply pallets/simple_democracy/src/lib.rs:125]: Yes

File: pallets/simple_democracy/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+        fn deposit_event() = default;
+
+        /// Proxy function to forked democracy pallet's `propose`
+        // TODO: Fix weight
+        #[weight = 0]
+        fn propose(origin, proposal_hash: T::Hash, #[compact] value: BalanceOf<T>) {
+            <pallet_democracy::Module<T>>::propose(origin, proposal_hash, value)?;
+        }
+
+        /// Proxy function to forked democracy pallet's `second`
```
[the remark this answers is not in the document]
Comment (lovesh) [reply pallets/simple_democracy/src/lib.rs:69]: Will change.

File: pallets/simple_democracy/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+        }
+
+        // TODO: Fix weight
+        #[weight = 0]
+        fn fast_track(origin, proposal_hash: T::Hash, voting_period: T::BlockNumber, delay: T::BlockNumber) {
+            <pallet_democracy::Module<T>>::fast_track(origin, proposal_hash, voting_period, delay)?;
+        }
+
+        /// Almost similar to forked democracy pallet's `vote` with the difference that vote balances
+        /// are disregarded and split voting is not allowed
```
[the remark this answers is not in the document]
Comment (lovesh) [reply pallets/simple_democracy/src/lib.rs:90]: I was trying to keep the interface same as the forked pallets but I guess I don't need to. I can change

File: pallets/simple_democracy/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+            }
+            Ok(())
+        } else {
+            Err(Error::<T>::OnlyStandardVotingAllowed.into())
+        }
+    }
+
+    /// Add vote to tally
+    fn add_to_tally(
+        vote: AccountVote<BalanceOf<T>>,
```
[the remark this answers is not in the document]
Comment (lovesh) [reply pallets/simple_democracy/src/lib.rs:251]: Yes

File: pallets/simple_democracy/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+    }
+
+    /// Get all deposits of a public proposal
+    pub fn deposit_of(prop_index: PropIndex) -> Option<(Vec<T::AccountId>, BalanceOf<T>)> {
+        <pallet_democracy::DepositOf<T>>::get(prop_index)
+    }
+}
+
+/// Transfer slashes to Treasury account defined in the PoA module
+impl<T: Trait> OnUnbalanced<NegativeImbalanceOf<T>> for Module<T> {
```
[the remark this answers is not in the document]
Comment (lovesh) [reply pallets/simple_democracy/src/lib.rs:315]: I see the point about decoupling but I wouldn't have now to do it. So keeping it as it is.

File: pallets/simple_democracy/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+    pub fn get_preimage(
+        image: T::Hash,
+    ) -> Option<PreimageStatus<T::AccountId, BalanceOf<T>, T::BlockNumber>> {
+        <pallet_democracy::Preimages<T>>::get(image)
+    }
+
+    /// Get all deposits of a public proposal
+    pub fn deposit_of(prop_index: PropIndex) -> Option<(Vec<T::AccountId>, BalanceOf<T>)> {
+        <pallet_democracy::DepositOf<T>>::get(prop_index)
+    }
```
[the remark this answers is not in the document]
Comment (lovesh) [reply pallets/simple_democracy/src/lib.rs:311]: They were meant to potentially expose as RPC so keeping them for now.

File: pallets/simple_democracy/src/tests.rs [hunk 11]
```diff
@@ -0,0 +1,1241 @@
...
+        assert_eq!(Democracy::public_props().len(), 1);
+
+        // Proposer's and backers' free balance decreases and that balance is reserved.
+        assert_eq!(Balances::reserved_balance(proposer), deposit);
+        assert_eq!(Balances::reserved_balance(backer_1), deposit);
+        assert_eq!(Balances::reserved_balance(backer_1), deposit);
+        assert_eq!(Democracy::deposit_of(0).unwrap().0.len(), 3);
+        assert_eq!(Democracy::deposit_of(0).unwrap().1, deposit);
+
+        let treasury_balance = PoAModule::treasury_balance();
```
[the remark this answers is not in the document]
Comment (lovesh) [reply pallets/simple_democracy/src/tests.rs:796]: Running out of time here so can't now. But we catch this imbalance while slashing

File: runtime/src/lib.rs [hunk 12]
```diff
@@ -389,6 +431,162 @@ impl anchor::Trait for Runtime {
...
+    /*type OperationalPreimageOrigin = EnsureOneOf<AccountId,
+        pallet_collective::EnsureMember<AccountId, CouncilCollective>,
+        pallet_collective::EnsureMember<AccountId, TechnicalCollective>,
+    >;*/
+    /*type VetoOrigin = EnsureOneOf<
+        AccountId,
+        EnsureRoot<AccountId>,
+        pallet_collective::EnsureProportionMoreThan<_1, _1, AccountId, TechnicalCollective>,
+        Success=AccountId
+    >;*/
```
[the remark this answers is not in the document]
Comment (lovesh) [reply runtime/src/lib.rs:581]: Will remove

File: runtime/src/lib.rs [hunk 13]
```diff
@@ -614,6 +819,12 @@ impl_runtime_apis! {
             add_benchmark!(params, batches, token_migration, MigrationModule);
             add_benchmark!(params, batches, frame_system, SystemBench::<Runtime>);
 
+            add_benchmark!(params, batches, pallet_collective, Council);
+            add_benchmark!(params, batches, pallet_democracy, Democracy);
+            add_benchmark!(params, batches, pallet_scheduler, Scheduler);
+
+            // TODO: Add benchmark for democracy as well
```
[the remark this answers is not in the document]
Comment (lovesh) [reply runtime/src/lib.rs:826]: Yes

File: runtime/src/lib.rs [hunk 14]
```diff
@@ -389,6 +431,162 @@ impl anchor::Trait for Runtime {
...
+    type Event = Event;
+    type Currency = Balances;
+    type EnactmentPeriod = EnactmentPeriod;
+    type LaunchPeriod = LaunchPeriod;
+    type VotingPeriod = VotingPeriod;
+    type CooloffPeriod = CooloffPeriod;
+    type MinimumDeposit = MinimumDeposit;
+    type ExternalOrigin = CouncilMember;
+    type ExternalMajorityOrigin = CouncilMember;
+    type ExternalDefaultOrigin = RootOrMoreThanHalfCouncil;
```
[the remark this answers is not in the document]
Comment (lovesh) [reply runtime/src/lib.rs:554]: This is required like `ExternalOrigin` just to compile, we can't set it to `()`

File: runtime/src/lib.rs [hunk 15]
```diff
@@ -389,6 +431,162 @@ impl anchor::Trait for Runtime {
...
+impl pallet_democracy::Trait for Runtime {
+    type Proposal = Call;
+    type Event = Event;
+    type Currency = Balances;
+    type EnactmentPeriod = EnactmentPeriod;
+    type LaunchPeriod = LaunchPeriod;
+    type VotingPeriod = VotingPeriod;
+    type CooloffPeriod = CooloffPeriod;
+    type MinimumDeposit = MinimumDeposit;
+    type ExternalOrigin = CouncilMember;
```
[the remark this answers is not in the document]
Comment (lovesh) [reply runtime/src/lib.rs:552]: Same as above

File: pallets/simple_democracy/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (lovesh) [reply pallets/simple_democracy/src/lib.rs:11]: Done

File: pallets/simple_democracy/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (lovesh) [reply pallets/simple_democracy/src/lib.rs]: Done

File: pallets/simple_democracy/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (lovesh) [reply pallets/simple_democracy/src/lib.rs]: Done

File: pallets/simple_democracy/src/lib.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (lovesh) [reply pallets/simple_democracy/src/lib.rs:154]: Changed

File: pallets/simple_democracy/src/lib.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (lovesh) [reply pallets/simple_democracy/src/lib.rs:387]: Changed

File: runtime/src/lib.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (lovesh) [reply runtime/src/lib.rs:520]: Added a note

File: runtime/src/lib.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (lovesh) [reply runtime/src/lib.rs:523]: Added a note

File: pallets/simple_democracy/src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (lovesh) [reply pallets/simple_democracy/src/lib.rs]: Done

File: runtime/src/lib.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (lovesh) [reply runtime/src/lib.rs]: Done

File: runtime/src/lib.rs [hunk 16]
```diff
@@ -389,6 +398,190 @@ impl anchor::Trait for Runtime {
...
+            | Call::PoAModule(_)
+            | Call::DIDModule(_)
+            | Call::Revoke(_)
+            | Call::BlobStore(_)
+            | Call::Anchor(_)
+            | Call::Master(_)
+            | Call::Sudo(_)
+            | Call::MigrationModule(_)
+            | Call::RandomnessCollectiveFlip(_) => true,
+            _ => false,
```
Comment (bddap) [inline runtime/src/lib.rs:579]: Please match exhaustively on `Call`. By that I mean don't use the `_` as a fall-through.

```suggestion
            Call::ForkedDemocracy(_) => false,
```

Comment (lovesh) [reply runtime/src/lib.rs]: Made the change. Is that ok?

**Review by bddap — APPROVED** [review]

**Pull request merged**